derrida/books/management/commands/instance_data.py

```diff
@@ -0,0 +1,122 @@
+'''
+Manage command to export instance data.
+
+Takes an optional argument to specify the output directory. Otherwise,
+files are created in the current directory.
+'''
+
+import codecs
+from collections import OrderedDict
+import csv
+import json
+import os.path
+
+from django.core.management.base import BaseCommand
+
+from derrida.books.models import Instance
+
+
+class Command(BaseCommand):
+    '''Export reference data for each Derrida Work as CSV and JSON'''
+    help = __doc__
+
+    #: fields for CSV output
+    csv_fields = [
+        'id', 'item_type', 'work_title', 'work_short_title',
+        'alternate_title', 'work_year', 'copyright_year',
+        # TODO: take into account print date day/month/year known fields
+        # 'print_date', 
+        'work_authors', 'publisher', 'pub_place',
+        'is_extant', 'is_annotated', 'is_translation', 'has_dedication',
+        'has_insertions', 'copy', 'dimensions', 'work_uri',
+        'work_subjects', 'languages', 'journal_title',
+        'book_title', 'book_title_uri', 
+        'start_page', 'end_page',
+        'has_digital_edition', 'uri', 'zotero_id'
+    ]
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-d', '--directory',
+            help='Specify the directory where files should be generated')
+
+    def handle(self, *args, **kwargs):
+        base_filename = 'derrida-instance-data'
+        if kwargs['directory']:
+            base_filename = os.path.join(kwargs['directory'], base_filename)
+
+        instancedata = [self.instance_data(instance) for instance in Instance.objects.filter(cited_in__isnull=False)]
+
+        # list of dictionaries can be output as is for JSON export
+        with open('{}.json'.format(base_filename), 'w') as jsonfile:
+            json_instancedata = [{field: inst[field] for field in inst.keys() if inst[field] not in [None, '', []]} for inst in instancedata]
+            json.dump(json_instancedata, jsonfile, indent=2)
+
+        # generate CSV export
+        with open('{}.csv'.format(base_filename), 'w') as csvfile:
+            # write utf-8 byte order mark at the beginning of the file
+            csvfile.write(codecs.BOM_UTF8.decode())
+
+            csvwriter = csv.DictWriter(csvfile, fieldnames=self.csv_fields)
+            csvwriter.writeheader()
+
+            for instance in instancedata:
+                csvwriter.writerow(self.flatten_dict(instance))
+
+    def instance_data(self, instance):
+        '''Generate a dictionary of data to export for a single
+         :class:`~derrida.books.models.Instance` object'''
+        return OrderedDict([
+            ('id', instance.get_uri()),
```

Comment:
- [ ] Check if URIs are numbers or slugs in existing dataset

Reply:
@rlskoeser The URIs are numbers in the published dataset (e.g. https://derridas-margins.princeton.edu/titles/143/) Let me know what next steps would look like for dealing with this

Reply:
After looking at it a bit more, I suspected they might be. I wonder if I did that because I thought they'd be more stable than the slug urls?

I don't think we should change them from what were in the published datasets... but we'll need to make sure we handle them in the archive site. I don't know if warcs handle redirects or not, we can test that. That would be simplest — we can use the list of URIs from the datasets as input to the crawler to make sure they are included. If not, we may have to have a list of nginx redirects — but from what I can tell that's still completely fine at the scale we need. (Thinking about it more, the nginx redirect is probably faster than the warc playback even if it's supported — so maybe we just plan to go that route.)

Reply:
Great, should I make that an Asana task or track that somewhere?

Reply:
how about an asana subtask for the archive site? ensure we handle all uris in published datasets (we'll want to make sure they all work no matter what they are)

Reply:
could add a separate task to setup redirects for title uris (and/or check redircts in the warc)